### PR TITLE
fixing timeouts for ackable messages

### LINKED ISSUE #745: Timeout when add to group
Hi, 
We have a Web API (hosted in azure using app service plan) configured with ASRS. At times, we are experiencing timeouts when add connection into a group. Here is our hub method to add to group.

public async Task AddToGroup(string groupName)
        {
            await Groups.AddToGroupAsync(Context.ConnectionId, groupName);
        }

Here is the exception message 
"Ack-able message Microsoft.Azure.SignalR.Protocol.JoinGroupWithAckMessage waiting for ack timed out." and trace
System.TimeoutException:
   at Microsoft.Azure.SignalR.ServiceConnectionContainerBase+<WriteAckableMessageAsync>d__49.MoveNext (Microsoft.Azure.SignalR.Common, Version=1.0.14.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.SignalR.MultiEndpointServiceConnectionContainer+<WriteAckableMessageAsync>d__18.MoveNext (Microsoft.Azure.SignalR.Common, Version=1.0.14.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at BDL.TwilioAPI.Hub.CallStatusHub+<AddToGroup>d__0.MoveNext (TwilioAPI, Version=1.0.0.0, Culture=neutral, PublicKeyToken=nullTwilioAPI, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: D:\a\1\s\TwilioAPI\Hub\CallStatusHub.csTwilioAPI, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: 9)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1+<ExecuteHubMethod>d__15.MoveNext (Microsoft.AspNetCore.SignalR.Core, Version=1.0.4.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1+<Invoke>d__13.MoveNext (Microsoft.AspNetCore.SignalR.Core, Version=1.0.4.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)

What would be the cause and how can we fix this? Our clients facing much of inconvenience in retrying. The ASRS is configured with "Classic" mode and Standard pricing tier. The ASRS metrics were alright in terms  on server and client connections when these timeout occurs.

We are using azure-signalr 1.0.14 version.

What would be the cause for this issue and is there any workaround that can help us?

Regards,
Tresa

### LINKED ISSUE #1225: Ack-able message Microsoft.Azure.SignalR.Protocol.JoinGroupWithAckMessage waiting for ack timed out.
### Describe the bug
when adding a user to a group I get this error:

[2021-02-22 08:37:04Z WRN] Failed to send message null.
System.TimeoutException: Ack-able message Microsoft.Azure.SignalR.Protocol.JoinGroupWithAckMessage waiting for ack timed out.
at Microsoft.Azure.SignalR.ServiceConnectionContainerBase.WriteAckableMessageAsync(ServiceMessage serviceMessage, CancellationToken cancellationToken)
at Microsoft.Azure.SignalR.MultiEndpointMessageWriter.<>c__DisplayClass8_0.<<WriteAckableMessageAsync>b__0>d.MoveNext()
--- End of stack trace from previous location ---
at Microsoft.Azure.SignalR.MultiEndpointMessageWriter.<>c__DisplayClass9_0.<<WriteMultiEndpointMessageAsync>b__2>d.MoveNext()
--- End of stack trace from previous location ---
at Microsoft.Azure.SignalR.MultiEndpointMessageWriter.WriteAckableMessageAsync(ServiceMessage serviceMessage, CancellationToken cancellationToken)
at Microsoft.Azure.SignalR.ServiceLifetimeManagerBase`1.WriteCoreAsync[T](T message, Func`2 task)

### To Reproduce
 call await Groups.AddToGroupAsync(Context.ConnectionId, "groupid", Context.ConnectionAborted); in OnConnectedAsync() in a hub. Occurs most of the time, but not always. Happens even with barely any server load.

### Further technical details
Microsoft.Azure.SignalR 1.7.1 in .NET 5 (linux docker container)

### PR DESCRIPTION
**Pull request by KKhurin**
### This fixes the longstanding problem with sporadically having ack timeouts when adding connections to groups.
 - the culprit behind the timeouts is a race condition when sending service message to several endpoints simultaneously which may lead to overriding AckId of a message that has not been sent yet
 - the stickiness to the service connection (introduced in SDK 1.7.1) makes this race condition happen a lot more often due to increased likelihood of going async (and increased chance of the race condition) because many scenarios would use the same connection with higher concurrency.

Fixes #745
Fixes #1225

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs [hunk 1]
```diff
@@ -94,10 +94,21 @@ internal void ReadExtensionMembers(ref MessagePackReader reader)
...
+        /// The purpose of Clone is to make a copy of everything that may get modified throughout the lifetime of the message
+        /// </summary>
+        ServiceMessage Clone { get; }
+    }
+
     /// <summary>
     /// Interface of ack-able message 
     /// </summary>
-    public interface IAckableMessage
+    public interface IAckableMessage : IClonableServiceMessage
```
[the remark this answers is not in the document]
Comment (KKhurin) [reply src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs:111]: :+1:
It indeed looks like the shallow member-wise clone is all we need! We can make `ServiceMessage.Clone` virtual if any message needs to cone itself differently in the future.

**Comment by vicancy** [comment]
Quote @vwxyzh 's proposal below to avoid changing the protocol

Change https://github.com/Azure/azure-signalr/blob/dev/src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs#L245 to

```cs

RawMessage raw = null;

lock (ackableMessage)

{

  ackableMessage.AckId = id;


  
  raw = WriteRawMessage(ackableMessage);

  }
```

**Comment by KKhurin** [comment]
> Quote @vwxyzh 's proposal below which avoids changing the protocol
> 
> Change https://github.com/Azure/azure-signalr/blob/dev/src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs#L245 to
> 
> ```cs
> RawMessage raw = null;
> 
> lock (ackableMessage)
> {
>   ackableMessage.AckId = id;
>   raw = WriteRawMessage(ackableMessage);
> }
> ```
Didn't originally choose this because of:
1. it allocates bytes for the raw message much earlier than it currently does - e.g. increases memory pressure (although making a copy also adds an extra allocation which by its nature should be smaller than the serialized one).
2. introduces a lock which is arguably heavier than the shallow member-wise clone.

I like the simplicity of @[yzt](https://github.com/Y-Sindo)'s approach if we can accept the new member in ServiceMessage.

File: src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs [hunk 2]
```diff
@@ -12,6 +12,11 @@ namespace Microsoft.Azure.SignalR.Protocol
     /// </summary>
     public abstract class ServiceMessage
     {
+        /// <summary>
+        /// Clone should make a copy of everything that may get modified throughout the lifetime of the message
+        /// The default implementation is a shallow copy as it fits the current needs.
+        /// </summary>
+        public virtual ServiceMessage Clone { get => MemberwiseClone() as ServiceMessage; }
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs:19]: use method `Clone()` instead?

File: src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs [hunk 3]
```diff
@@ -16,7 +16,7 @@ public abstract class ServiceMessage
         /// Clone should make a copy of everything that may get modified throughout the lifetime of the message
         /// The default implementation is a shallow copy as it fits the current needs.
         /// </summary>
-        public virtual ServiceMessage Clone { get => MemberwiseClone() as ServiceMessage; }
+        public virtual ServiceMessage Clone () => MemberwiseClone() as ServiceMessage;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs:19]: nit: extra space

**Review by vicancy — APPROVED** [review]

**Pull request merged**